### PR TITLE
Closes #2543 Preload for fonts

### LINKED ISSUE #2543: Preload For Fonts
## What are we proposing to do? 

Font files that are requested in a CSS file or later in the page load can be downloaded faster by preloading them. With this update, we will provide the necessary tools for our users to accomplish this. 

## Why are we doing this? 

It is common practice to reference font files in CSS files. In this case, the browser has to download and parse the CSS files before it can download the fonts. Modern browsers can download multiple files in parallel. 

With this feature, our users will be able to “tell” the browser about these hidden fonts. The feature also informs the browser that files are indeed fonts. Modern browsers handle fonts very well 

This feature will also satisfy the “Preload key requests” suggestion of PageSpeed. 

image4

## Benchmark Tests

Jonathan did benchmarks for this feature on https://wp-rocket.me/ with positive results. Here is the waterfall

**No Preload:** 

no-preload_LI
Full scale: https://jmp.sh/A336ebO

**With Preload:** 

with-font-preload_LI
Full scale: https://jmp.sh/T8Gb6Ko

Note how the fonts take considerably lower time to load. 

Another example is depicted here: https://developers.google.com/web/tools/lighthouse/audits/preload
User expectations and experience. How will we do this feature?

## UI Proposal

Feature location: **Preload tab**
Section title: **Preload Fonts**
Section description: **Improves performance by helping browsers discover fonts in CSS files. More info** (Thanks Lucy https://github.com/wp-media/wp-rocket/issues/2543#issuecomment-615335738)

Text area title: **Fonts to preload**
Text area description: **Specify urls of the font files to be preloaded (one per line).**
Text area guideline (in green): **The domain part of the URL will be stripped automatically.
Allowed font extensions: ttf, otf, woff, woff2, svg.**

preload fonts

**A related change is also recommended for cache preloading**. Currently the section title simply says “Preload”. To avoid any confusion, let’s rename it to “**Preload Cache**”.

image1

Also update the tab description from “Generate cache files” to: **Generate cache files, preload fonts**

image6

It will look so: 

image3

## Technical Specs

**The following is the expected output format in the <head> of the website**

`<link rel="preload" as="font" href="https://domain.com/wp-content/themes/theme/assets/fonts/font-file.woff" crossorigin>`

See example in the source code: https://wp-rocket.me/?nocache
(crossorigin is required, [reference](https://www.smashingmagazine.com/2016/02/preload-what-is-it-good-for/#early-loading-of-fonts))

**Preload for Google fonts:** 

- In every test conducted, PageSpeed never recommended to preload Google fonts.
- Even in tests where the Google font was manually specified in the same stylesheet with a local font, PageSpeed only recommended to preload the local font. (https://jmp.sh/Bp2oukP, https://jmp.sh/ORy9W34)
- In a real use case, Google fonts are added to the head, so they are discovered early and shouldn’t have an impact. 
- However, if we get user feedback or we discover new information about the same, we can consider this in a future iteration. 
- If we end up doing this ^ in the future, then the user would enter the complete font url (example: https://fonts.gstatic.com/s/manrope/v1/xn7_YHE41ni1AdIRqAuZuw1Bx9mbZk59FN_G-bnBeA.woff2) in the WP Rocket UI. 

**User Input**

- During sanitization, strip off the domain part of the url (just as we do for the text areas in the File Optimization tab)
- During sanitization, check for allowed font formats( `ttf, otf, woff, woff2, svg` ) and remove any input that isn't this format. We can remove it silently because we already display allowed formats in the UI
- Add a placeholder (just like we have for dns prefetch https://jmp.sh/eiR9qun) in the text area that shows an example (`/wp-content/themes/theme/assets/fonts/font-file.woff`)

## Risks

The entire feature is opt-in and only adds meta tags. So it’s extremely low risk. At the very most, the user might enter incorrect fonts and that might cause a minor performance penalty.

## Sub-tasks
- [x] [UI](https://github.com/wp-media/wp-rocket/issues/2545)
- [x] [Option management](https://github.com/wp-media/wp-rocket/issues/2546)
- [x] [Front](https://github.com/wp-media/wp-rocket/issues/2547)

## Acceptance Criteria

- Enter a font url in the UI
- Font is added as `rel="preload" as="font"` link meta tag in the header
- Enter the url to the font with domain.com (http://domain.com/wp-content/themes/theme/assests/font.woff). Check if the domain part is stripped off. 
- Enter a font url with a non-allowed file extension (eg: .eot). The font url should be removed from the UI when saved. This should be added to the `<head>` for preload. 
- Enable CDN
- Font urls are rewritten with CDN CNAME

**What if?** 

- The font url entered is invalid? (404)
^ My recommendation is to do nothing for now. If this becomes a common thing, we can figure it out then. 

## Documentation and Translations

- A new doc describing the feature. Doc needs to educate the user on how to identify the font files (PageSpeed report could be a good place to find them). 
- When Optimize CSS delivery is enabled, since the CSS file is preloaded, PageSpeed will not suggest to preload the font file. To discover fonts via PageSpeed, it’s recommended to disabled Optimize CSS delivery and then run the test.

### PR DESCRIPTION
**Pull request by Tabrisrp**
Closes epic #2543

This PR contains all the changes for the feature and completes the epic. Each sub-task has been reviewed and QA already.

**Missing**
- [x] Preserve the query string & hash in the URL in the option field
- [x] An URL with the CDN URL should be preserved when adding it in the option field
- [x] Documentation FR links for Beacon
- [x] QA

### CODE REVIEW

**Comment by webtrainingwheels** [comment]
If the CDN option is activated, and the customer enters their font using their CDN url, eg.
`https://cdn.lucy.com/wp-content/themes/paperback/inc/fontawesome/fonts/fontawesome-webfont.woff2?v=4.4.0`

Currently we don't recognize it and remove the whole font url from the box when saving. I think it would be ideal to recognize the CDN CNAME and just remove the domain, as if it were the local domain.

**Comment by webtrainingwheels** [comment]
Some fonts will have a query string eg: `wp-content/themes/paperback/inc/fontawesome/fonts/fontawesome-webfont.woff2?v=4.4.0`
Because the query string is stripped off, this creates 2 issues:
-  there's now 2 requests for the font in the Waterfall- one with the query string and one without: https://jmp.sh/RuRlH75
-   Chrome shows this warning: https://jmp.sh/1eK19KA

**Comment by Tabrisrp** [comment]
@hellofromtonya @arunbasillal @GeekPress Please note the 2 comments above from Lucy.

> there's now 2 requests for the font in the Waterfall- one with the query string and one without: https://jmp.sh/RuRlH75

Is it only on GTmetrix, or also in the browser network?

> Chrome shows this warning

About this warning:
- it's also present in Safari

This means the font should be added to the preload field really only if it's used early in the page. I guess this will need to be in the doc about the feature?

**Comment by webtrainingwheels** [comment]
@Tabrisrp 
> This means the font should be added to the preload field really only if it's used early in the page. I guess this will need to be in the doc about the feature?

To be clear, the Chrome warning only shows due to the query string issue. 
Preloading the font file correctly with the query string removes the Chrome warning.

**Comment by Tabrisrp** [comment]
Thanks for the clarification 👍

**Comment by webtrainingwheels** [comment]
The EN link for documentation will be: https://docs.wp-rocket.me/article/1317-preload-fonts
The article ID is: 5eab7729042863474d19f647

**Comment by Tabrisrp** [comment]
PR is updated to fix the issues reported by Lucy above
- Preserve the query string/fragment when it's present in the original URL
- Preserve the entry if it uses an URL corresponding to one of the URL in the CDN cnames field

File: inc/Engine/Preload/Fonts.php [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		$fonts = (array) apply_filters( 'rocket_preload_fonts', $fonts );
+		$fonts = array_map( [ $this, 'sanitize_font' ], $fonts );
+		$fonts = array_filter( $fonts );
+
+		if ( empty( $fonts ) ) {
+			return;
+		}
+
+		$base_url = get_rocket_parse_url( home_url() );
+		$base_url = "{$base_url['scheme']}://{$base_url['host']}";
```
Comment (hellofromtonya) [inline inc/Engine/Preload/Fonts.php:99]: @Tabrisrp Do we get the base URL in other places in WPR? If yes, then I'm thinking we'll want to centralize this functionality to reuse everywhere. What do you think?

Comment (Tabrisrp) [reply inc/Engine/Preload/Fonts.php:99]: we have different need in different places in fact, sometimes we need only the `host`, sometimes the base URL, sometimes we need that but with the `site_url()` instead.

File: tests/Unit/inc/Engine/Admin/Settings/Settings/sanitizeCallback.php [hunk 2]
```diff
@@ -19,7 +19,7 @@ class Test_SanitizeCallback extends TestCase {
 	public function setUp() {
 		parent::setUp();
 
-		$this->options  = $this->createMock( Options::class );
+		$this->options  = $this->createMock( Options_Data::class );
```
Comment (crystinutzaa) [inline tests/Unit/inc/Engine/Admin/Settings/Settings/sanitizeCallback.php:22]: This should be changed to Mockery:mock(), right?

Comment (hellofromtonya) [reply tests/Unit/inc/Engine/Admin/Settings/Settings/sanitizeCallback.php:22]: Yes it should.

Comment (Tabrisrp) [reply tests/Unit/inc/Engine/Admin/Settings/Settings/sanitizeCallback.php]: https://github.com/wp-media/wp-rocket/pull/2592/commits/3a497ea3a2dcefb601fc4fc0ab9c785ddd8046e7

**Comment by Tabrisrp** [comment]
@wp-media/dev-qa Ready for a new QA, fixing the 2 issues reported above by Lucy

**Comment by Tabrisrp** [comment]
> Output: https://jmp.sh/7mLayVB 👍
Output in Frontend: These were not added to the front end ❌ (@Tabrisrp please note)

I don't understand, your image shows the output on the frontend, with the expected link?

**Comment by hellofromtonya** [comment]
@wp-media/dev-qa This PR is ready for QA.

**Comment by Tabrisrp** [comment]
These should be fixed now too

**Comment by Tabrisrp** [comment]
@arunbasillal Edge case indeed, but I was able to change the code to accomodate that. Can you give it another try?

**Comment by arunbasillal** [comment]
@Tabrisrp Thanks for the update. 

Right now, when I input `https://example.com/wp-content/themes/font.woff`, and when the CDN is set to `https://example.com/path`, the input is removed after sanitization. 

Can you please try the exact same input at your end and see what happens?

**Comment by Tabrisrp** [comment]
Is `example.com` the site URL or a different URL? It's unclear.

If `https://example.com/path` is the CDN URL in the cnames, and you input `https://example.com/wp-content/themes/font.woff`, the input will be removed, because it doesn't match the CDN URL.

`example.com` !== `example.com/path`.

**Comment by arunbasillal** [comment]
@Tabrisrp `example.com` is the CDN CNAME and NOT the website URL. 

> If https://example.com/path is the CDN URL in the cnames, and you input https://example.com/wp-content/themes/font.woff, the input will be removed, because it doesn't match the CDN URL.
> example.com !== example.com/path.

This was the edge case reported. And this is indeed covered by the update. So we are good 👍 

However, I think we have created a regression. 

This is the input I used:
```
#Local fonts with domain (wp-rocket-dev.local) is the local domain
http://wp-rocket-dev.local/wp-content/themes/theme/font.otf
http://wp-rocket-dev.local/wp-content/themes/theme/font.ttf
http://wp-rocket-dev.local/wp-content/themes/theme/font.svg
http://wp-rocket-dev.local/wp-content/themes/theme/font.woff
http://wp-rocket-dev.local/wp-content/themes/theme/font.woff2
http://wp-rocket-dev.local/wp-content/themes/theme/font.eot

#External domain
http://domain.com/wp-content/themes/theme/font-external.woff

#No domain
/wp-content/themes/theme/font-no-domain.woff

#Disallowed extension
/wp-content/themes/theme/font.eot

#Query String
wp-content/themes/paperback/inc/fontawesome/fonts/fontawesome-webfont.woff2?v=4.4.0

#Query string with leading slash
/wp-content/themes/paperback/inc/fontawesome/fonts/fontawesome-webfont.woff2?v=4.4.0

#Fragment
wp-content/themes/paperback/inc/fontawesome/fonts/fontawesome-webfont.woff2#Fragment

#CDN with Path (CDN CNAME set to https://example.com/path)
https://example.com/path/wp-content/themes/paperback/inc/fontawesome/fonts/fontawesome-cdn-path.woff2?v=path

#CDN without Path
https://example.com/wp-content/themes/paperback/inc/fontawesome/fonts/fontawesome-cdn-no-path.woff2?v=path
```

Output was: 

```
/wp-content/themes/theme/font-no-domain.woff
/wp-content/themes/paperback/inc/fontawesome/fonts/fontawesome-webfont.woff2?v=4.4.0
/wp-content/themes/paperback/inc/fontawesome/fonts/fontawesome-webfont.woff2#Fragment
/wp-content/themes/paperback/inc/fontawesome/fonts/fontawesome-cdn-path.woff2?v=path
```

Looks like `http://wp-rocket-dev.local` (the local site url) was treated as external domain. 
Also it recorded this in debug.log - http://snippi.com/raw/uvvl4tc when you try the input exactly as above including the #Titles

**Comment by Tabrisrp** [comment]
The `#titles` should not be included in the input, using the `#` will make the sanitizer think it's part of a fragment URL.

I fixed the issue you noted above, it was a mistake from me in the logic

**Pull request merged**